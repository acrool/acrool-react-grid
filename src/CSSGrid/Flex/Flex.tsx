import React from 'react';
import styled from 'styled-components';
import {TStyledProps, IFlexProps} from '../../types';


/**
 * Flex Component
 */
const Flex = styled.div.attrs((props: TStyledProps<IFlexProps>) => ({
    'data-grid': 'flex',
}))`
    display: flex;
    flex-direction: row;
`;

export default Flex;

