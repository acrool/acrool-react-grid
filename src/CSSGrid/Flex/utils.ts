import {
    TStyledProps,
    IGridProps,
    TGridGap,
    TGridTemplate,
    TGridGaps,
    TGridCol,
    TSpan,
    IGridColProps, TFlexDirection, IFlexProps, TDirection
} from '../../types';
import {noXsMediaSizes} from '../../config';
import media from '../../media';
import {css} from 'styled-components';





interface ICSSGetter {
    direction: (direction: TDirection) => string

    gap: (gapNum: TGridGap) => string
    columnGap: (gapNum: TGridGap) => string
    rowGap: (gapNum: TGridGap) => string
}

export const cssGetter: ICSSGetter = {
    direction: (direction) => {
        return `flex-direction: ${direction};`;
    },
    gap: (gapNum) => {
        return `
        gap: ${gapNum};
        `;
    },
    columnGap: (gapNum) => {
        return `
        column-gap: ${gapNum};
        `;
    },
    rowGap: (gapNum) => {
        return `
        row-gap: ${gapNum};
        `;
    },
};



/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IFlexProps>) => {
    return noXsMediaSizes
        .map(sizeName => {
            const mediaColSpanProps = typeof props.direction === 'object' ? props.direction[sizeName]: undefined;
            const mediaGapProps = typeof props.gap === 'object' ? props.gap[sizeName]: undefined;
            const mediaColumnGapProps = typeof props.columnsGap === 'object' ? props.columnsGap[sizeName]: undefined;
            const mediaRowGapProps = typeof props.rowsGap === 'object' ? props.rowsGap[sizeName]: undefined;

            return media[sizeName]`

            ${typeof mediaColSpanProps !== 'undefined' && css`
                ${cssGetter.direction(mediaColSpanProps)};
            `}

            ${typeof mediaGapProps !== 'undefined' && css`
                ${cssGetter.gap(mediaGapProps)};
            `}
            ${typeof mediaColumnGapProps !== 'undefined' && css`
                ${cssGetter.columnGap(mediaColumnGapProps)};
            `}
            ${typeof mediaRowGapProps !== 'undefined' && css`
                ${cssGetter.rowGap(mediaRowGapProps)};
            `}
    `;
        });
};

