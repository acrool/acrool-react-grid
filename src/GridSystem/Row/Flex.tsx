import styled from 'styled-components';
import CSS from 'csstype';
import {TFlexDirection} from '../../types';


/**
 * Flex Component
 */
const Flex = styled.div<{
    style?: CSS.Properties,
    className?: string,
    direction?: TFlexDirection
}>`
  box-sizing: border-box;
  display: flex;
  flex-direction: ${props => props.direction ? props.direction : 'row'};
`;

export default Flex;
