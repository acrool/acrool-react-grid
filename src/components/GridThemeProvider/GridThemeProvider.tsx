import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle} from 'styled-components';

import {defaultGridTheme, themeName} from '../../config';
import {IGridSetting} from '../../types';


interface IProps {
    gridTheme: Partial<IGridSetting>;
    children: React.ReactNode;
}

const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    
`;
