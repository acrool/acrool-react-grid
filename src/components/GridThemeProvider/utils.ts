import {TStyledProps, IRowProps, NoXsMediaSize} from '../../types';
import {mediaSizes} from '../../config';
import media from '../../media';


const renderMarginStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .m-${sizeName}-${idx} {margin: var(--bear-spacer-${idx}) !important;}
        .mt-${sizeName}-${idx}, .my-${sizeName}-${idx} {margin-top: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .mr-${sizeName}-${idx}, .mx-${sizeName}-${idx} {margin-right: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .mb-${sizeName}-${idx}, .my-${sizeName}-${idx} {margin-bottom: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .ml-${sizeName}-${idx}, .mx-${sizeName}-${idx} {margin-left: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        `;
    }).join('');
};

const renderPaddingStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .p-${sizeName}-${idx} {margin: var(--bear-spacer-${idx}) !important;}
        .pt-${sizeName}-${idx}, .py-${sizeName}-${idx} {padding-top: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pr-${sizeName}-${idx}, .px-${sizeName}-${idx} {padding-right: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pb-${sizeName}-${idx}, .py-${sizeName}-${idx} {padding-bottom: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pl-${sizeName}-${idx}, .px-${sizeName}-${idx} {padding-left: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        `;
    }).join('');
};

const renderOrderStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .order-${sizeName}-${idx} {order: ${idx}!important;}
        `;
    }).join('');
};


/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IRowProps>) => {
    return mediaSizes
        .map(sizeName => {
            return media[sizeName]`
            .d-${sizeName}-none {display: none !important;}
            .d-${sizeName}-inline {display: inline !important;}
            .d-${sizeName}-inline-block {display: inline-block !important;}
            .d-${sizeName}-block {display: block !important;}
            .d-${sizeName}-flex {display: flex !important;}
            .d-${sizeName}-inline-flex {display: inline-flex !important;}
            .d-${sizeName}-table-cell {display: table-cell !important;}
            .d-${sizeName}-table {display: table !important;}
            .d-${sizeName}-table-row {display: table-row !important;}

            .flex-${sizeName}-row {flex-direction: row !important;}
            .flex-${sizeName}-column {flex-direction: column !important;}
            .flex-${sizeName}-row-reverse {flex-direction: row-reverse !important;}
            .flex-${sizeName}-column-reverse {flex-direction: column-reverse !important;}

            .flex-${sizeName}-wrap {flex-wrap: wrap !important;}
            .flex-${sizeName}-nowrap {flex-nowrap: nowrap !important;}
            .flex-${sizeName}-wrap-reverse {flex-nowrap: wrap-reverse !important;}

            .flex-${sizeName}-fill {flex: 1 1 auto !important;}
            .flex-${sizeName}-grow-0 {flex-grow: 0 !important;}
            .flex-${sizeName}-grow-1 {flex-grow: 1 !important;}
            .flex-${sizeName}-shrink-0 {flex-shrink: 0 !important;}
            .flex-${sizeName}-shrink-1 {flex-shrink: 1 !important;}

            .justify-content-${sizeName}-start {justify-content: flex-start !important;}
            .justify-content-${sizeName}-end {justify-content: flex-end !important;}
            .justify-content-${sizeName}-center {justify-content: flex-center !important;}
            .justify-content-${sizeName}-between {justify-content: space-between !important;}
            .justify-content-${sizeName}-around {justify-content: space-around !important;}

            .align-items-${sizeName}-start {align-items: flex-start !important;}
            .align-items-${sizeName}-end {align-items: flex-end !important;}
            .align-items-${sizeName}-center {align-items: center !important;}
            .align-items-${sizeName}-baseline {align-items: baseline !important;}
            .align-items-${sizeName}-stretch {align-items: stretch !important;}

            .align-content-${sizeName}-start {align-content: start !important;}
            .align-content-${sizeName}-end {align-content: flex-end !important;}
            .align-content-${sizeName}-center {align-content: center !important;}
            .align-content-${sizeName}-between {align-content: between !important;}
            .align-content-${sizeName}-around {align-content: around !important;}
            .align-content-${sizeName}-stretch {align-content: stretch !important;}

            .align-self-${sizeName}-start {align-self: flex-start !important;}
            .align-self-${sizeName}-end {align-self: flex-end !important;}
            .align-self-${sizeName}-center {align-self: center !important;}
            .align-self-${sizeName}-baseline {align-self: baseline !important;}
            .align-self-${sizeName}-stretch {align-self: stretch !important;}

            .text-${sizeName}-justify {text-align: justify !important;}
            .text-${sizeName}-left {text-align: left !important;}
            .text-${sizeName}-right {text-align: right !important;}
            .text-${sizeName}-center {text-align: center !important;}

            .m-${sizeName}-auto {margin: var(--bear-spacer-auto) !important;}
            .mt-${sizeName}-auto, .my-${sizeName}-auto {margin-top: auto !important;}
            .mr-${sizeName}-auto, .mx-${sizeName}-auto {margin-right: auto !important;}
            .mb-${sizeName}-auto, .my-${sizeName}-auto {margin-bottom: auto !important;}
            .ml-${sizeName}-auto, .mx-${sizeName}-auto {margin-left: auto !important;}

            ${renderMarginStyle(sizeName)}
            ${renderPaddingStyle(sizeName)}
            ${renderOrderStyle(sizeName)}

            .overflow-x-${sizeName}-hidden {overflow-x: hidden!important;}
            .overflow-y-${sizeName}-hidden {overflow-y: hidden!important;}
            .overflow-x-${sizeName}-auto {overflow-x: auto!important;}
            .overflow-y-${sizeName}-auto {overflow-y: auto!important;}
        `;
        });
};

