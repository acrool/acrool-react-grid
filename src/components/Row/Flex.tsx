import styled from 'styled-components';
import CSS from 'csstype';
import {TRowVertical, TRowHorizontal, TFlexDirection} from '../../types';


/**
 * Flex Component
 */
const Flex = styled.div<{
    style?: CSS.Properties,
    className?: string,
    direction?: TFlexDirection
    vertical?: TRowVertical; // 垂直
    horizontal?: TRowHorizontal; // 水平
}>`
  box-sizing: border-box;
  display: flex;
  flex-direction: ${props => props.direction ? props.direction : 'row'};
  justify-content: ${props => props.horizontal ? props.horizontal : 'left'};
  align-items: ${props => props.vertical ? props.vertical : 'top'};
`;

export default Flex;
