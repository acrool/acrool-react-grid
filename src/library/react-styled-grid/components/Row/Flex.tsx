import styled from 'styled-components/macro';
import {EAlign} from './types';

export enum EDirection {
    row='row',
    column ='column'
}

const Flex = styled.div<{
    direction?: EDirection
    vertical?: EAlign; // 垂直
    horizontal?: EAlign; // 水平
}>`
  box-sizing: border-box;
  display: flex;
  flex-direction: ${props => props.direction ?? EDirection.row};
  justify-content: ${props => props.horizontal ?? EAlign.start};
  align-items: ${props => props.vertical ?? EAlign.start};
`;

export default Flex;
