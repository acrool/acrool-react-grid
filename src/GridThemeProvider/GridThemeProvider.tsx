import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle, css} from 'styled-components';

import {defaultGridTheme, themeName} from '../config';
import {TStyledProps, TThemeProps, IGridThemeProviderProps} from '../types';
import {generateRWDStyled} from './utils';
import '../styles.css';



const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IGridThemeProviderProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    :root{
        --bear-spacer: 1rem;
        --bear-spacer-1: calc(1rem * .25);
        --bear-spacer-2: calc(1rem * .5);
        --bear-spacer-3: calc(1rem);
        --bear-spacer-4: calc(1rem * 1.5);
        --bear-spacer-5: calc(1rem * 3);
    }

    ${(props: TStyledProps<TThemeProps>) => css`
        ${generateRWDStyled(props)}
    `}

    [data-row][data-horizontal="left"]{
        justify-content: flex-start;
    }
    [data-row][data-horizontal="center"]{
        justify-content: center;
    }
    [data-row][data-horizontal="right"]{
        justify-content: flex-end;
    }
    [data-row][data-vertical="top"]{
        justify-content: flex-start;
    }
    [data-row][data-vertical="center"]{
        justify-content: center;
    }
    [data-row][data-vertical="bottom"]{
        justify-content: flex-end;
    }
`;
