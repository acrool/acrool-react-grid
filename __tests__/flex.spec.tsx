import React from 'react';
import 'jest-styled-components';
import {render} from '@testing-library/react';
import '@testing-library/dom';
import GridThemeProvider from '../src/GridThemeProvider';
import Flex from '../src/CSSGrid/Flex';

describe('Flex', () => {
    it('renders flex', () => {
        const {container} = render(<Flex>flex</Flex>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });
    it('renders flex column', () => {
        const {container} = render(<Flex column>flex</Flex>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });
    it('renders flex row', () => {
        const {container} = render(<Flex col="row">flex</Flex>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders flex column + md row', () => {
        const {container} = render(<Flex column md="row">flex</Flex>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

});
