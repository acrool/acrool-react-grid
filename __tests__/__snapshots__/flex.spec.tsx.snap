// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex renders flex 1`] = `
.c0 {
  display: flex;
}

<div
  class="c0"
>
  flex
</div>
`;

exports[`Flex renders flex column + md row 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
}

<div
  class="c0"
  data-grid="flex-column"
  md="row"
>
  flex
</div>
`;

exports[`Flex renders flex column 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
}

<div
  class="c0"
  data-grid="flex-column"
>
  flex
</div>
`;

exports[`Flex renders flex row 1`] = `
.c0 {
  display: flex;
}

<div
  class="c0"
  col="row"
>
  flex
</div>
`;
